### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by TheFakeOG**
Added a lot of new tests

### CODE REVIEW

File: GATEWAY/src/pages/login/login.ts [hunk 1]
```diff
@@ -25,6 +25,7 @@ export class LoginPage {
    * @param {string} user - username
    * @param {string} host - api host address
    * @param {number} port - mqtt port number
+   * @param {boolean} remember - if the login credentials should be remembered
```
Comment (nornes) [inline GATEWAY/src/pages/login/login.ts:28]: why is this added?

File: GATEWAY/src/providers/ble.service.ts [hunk 2]
```diff
@@ -22,11 +22,11 @@ export class BleService {
   };
 
   constructor(private events: Events,
-              private ble: BLE,
-              private devicesService: DevicesService,
-              private mqttClient: MqttClient,
-              private tilesApi: TilesApi,
-              private utils: UtilsService) {}
+              public ble: BLE,
```
Comment (nornes) [inline GATEWAY/src/providers/ble.service.ts:25]: why are you changing these to public?

**Review by aaursand — DISMISSED** [review]

File: GATEWAY/src/providers/tilesApi.service.ts [hunk 3]
```diff
@@ -12,12 +12,25 @@ export class TilesApi {
...
   private activeApp: Application;
+  flagThen: boolean = false;
+  flagCatch: boolean = false;
 
   constructor(private http: Http,
               private storage: Storage) {
   }
 
   /**
+   * The following code will mailny be used for getting private parameters
```
Comment (nornes) [inline GATEWAY/src/providers/tilesApi.service.ts]: again, same as what i said to the new methods in mqtt

**Review by nornes — DISMISSED** [review]

File: GATEWAY/src/pages/login/login.ts (hunk 1 above)
Comment (TheFakeOG) [reply GATEWAY/src/pages/login/login.ts:28]: Missing from the parameter description
It is used as a parameter in the connectToServer method, but was not described

File: GATEWAY/src/providers/ble.service.ts (hunk 2 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/ble.service.ts:25]: Testing purposes
Need to be able to get them one way or another
Can implement a get-method for these variables instead if that is better

File: GATEWAY/src/providers/tilesApi.service.ts (hunk 3 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/tilesApi.service.ts]: You mean ble?
If so, it is for the same reason
I need to be able to get these variables to properly test the class

Comment (nornes) [reply GATEWAY/src/providers/tilesApi.service.ts]: I forgot to publish the comment to the review -.- I said that i dont think those methods should be there, especially the getTilesApi and setClient in mqtt and that Id rather have the code that needed them untested if there is no other way to test it.

Comment (TheFakeOG) [reply GATEWAY/src/providers/tilesApi.service.ts]: Untested?

Comment (TheFakeOG) [reply GATEWAY/src/providers/tilesApi.service.ts]: Ok, I understand what you mean now, but it is unfortunately not an option, as we should aim for full unit test coverage

**Comment by TheFakeOG** [comment]
Well this is rather anoying...
Travis fails my PR even though all the tests passes locally
I do not know what i can do to make this work, so in the mean time if the code gets reviewed, i will merge it
Looks like it has something to do with the `karma start ./karma.conf.js --single-run` command, and not the tests themselves

File: GATEWAY/src/pages/login/login.ts (hunk 1 above)
Comment (nornes) [reply GATEWAY/src/pages/login/login.ts:28]: didn't see that it was inside a comment, good catch 👍

File: GATEWAY/src/providers/ble.service.ts [hunk 4]
```diff
@@ -29,6 +29,32 @@ export class BleService {
               private utils: UtilsService) {}
 
   /**
+   * The following code will mainly be used for getting private parameters
```
Comment (nornes) [inline GATEWAY/src/providers/ble.service.ts]: I still higly disapprove of these methods

File: GATEWAY/src/providers/devices.service.ts [hunk 5]
```diff
@@ -19,6 +19,9 @@ export class DevicesService {
   public getDevices = (): Device[] => {
     return this.devices;
   }
+  public setDevices = (dev: Device[]): void => {
```
Comment (nornes) [inline GATEWAY/src/providers/devices.service.ts]: to set new devices call newdevice several times instead. This method can be confusing for the developers who are taking over the project as it should never be used by non-testing code.

File: GATEWAY/src/providers/devices.service.ts [hunk 6]
```diff
@@ -68,6 +71,6 @@ export class DevicesService {
    */
   public clearDisconnectedDevices = (): void => { // TODO: Change name?
     const currentTime = (new Date()).getTime();
-    this.devices = this.devices.filter(device => device.lastDiscovered - currentTime < 60000);
+    this.devices = this.devices.filter(device => currentTime - device.lastDiscovered < 60000);
```
Comment (nornes) [inline GATEWAY/src/providers/devices.service.ts:64]: smart you are 👍

File: GATEWAY/src/providers/mqttClient.ts [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import { Injectable } from '@angular/core';
 import { BackgroundFetch } from '@ionic-native/background-fetch';
 import { Events } from 'ionic-angular';
-import mqtt from 'mqtt';
+import * as mqtt from 'mqtt';
```
Comment (nornes) [inline GATEWAY/src/providers/mqttClient.ts:4]: why? it worked before?

File: GATEWAY/src/providers/mqttClient.ts [hunk 8]
```diff
@@ -34,6 +34,26 @@ export class MqttClient {
   }
 
   /**
+   * The following code will mainly be used for getting private parameters
```
Comment (nornes) [inline GATEWAY/src/providers/mqttClient.ts]: again, strongly disapprove

File: GATEWAY/src/providers/tilesApi.service.ts [hunk 9]
```diff
@@ -8,6 +8,8 @@ import { Application, LoginData, VirtualTile } from './utils.service';
 
 @Injectable()
 export class TilesApi {
+  public flagThen: boolean = false;
```
Comment (nornes) [inline GATEWAY/src/providers/tilesApi.service.ts:11]: why?

File: GATEWAY/src/providers/tilesApi.service.ts [hunk 10]
```diff
@@ -18,6 +20,17 @@ export class TilesApi {
   }
 
   /**
+   * The following code will mainly be used for getting private parameters
```
Comment (nornes) [inline GATEWAY/src/providers/tilesApi.service.ts]: again, dissapproval

**Review by nornes — DISMISSED** [review]
still think the same about the changes. As for the test command it works in the current master so it should be fixed in this branch before it is merged

File: GATEWAY/src/providers/ble.service.ts (hunk 4 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/ble.service.ts]: I kno, fixing it now 👍

File: GATEWAY/src/providers/devices.service.ts (hunk 5 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/devices.service.ts]: That is not good testing etiquette.
I can explain it in detail if you want, but running `newDevice` several times is not an option

File: GATEWAY/src/providers/devices.service.ts (hunk 6 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/devices.service.ts:64]: Are you sarcastic?
I honestly can't tell since this comment is under change requests.
Even if you are or not, it did not work before, so i fixed it 😛

File: GATEWAY/src/providers/mqttClient.ts (hunk 7 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/mqttClient.ts:4]: Does not work when i try to test it
This version of the import works, and does not influence the code in any way
Just for testing purposes

File: GATEWAY/src/providers/mqttClient.ts (hunk 8 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/mqttClient.ts]: I kno, fixing it now :+1:

File: GATEWAY/src/providers/tilesApi.service.ts (hunk 9 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/tilesApi.service.ts:11]: Because I am not able to test the code without these flags
I can explain this in detail later, but for now it is vital for the testing

File: GATEWAY/src/providers/tilesApi.service.ts (hunk 10 above)
Comment (TheFakeOG) [reply GATEWAY/src/providers/tilesApi.service.ts]: I kno, fixing it now :+1:

**Comment by TheFakeOG** [comment]
Travis still does nt want to pass the tests, even though they pass...

File: GATEWAY/src/providers/devices.service.ts (hunk 6 above)
Comment (nornes) [reply GATEWAY/src/providers/devices.service.ts:64]: not sarcastic

File: GATEWAY/src/providers/mqttClient.ts (hunk 7 above)
Comment (nornes) [reply GATEWAY/src/providers/mqttClient.ts:4]: was just curious

File: GATEWAY/src/providers/devices.service.ts (hunk 6 above)
Comment (nornes) [reply GATEWAY/src/providers/devices.service.ts:64]: just laughing at my own stupidity when I wrote it

**Review by nornes — COMMENTED** [review]
Now we're talking :D good job! 👍

**Review by nornes — APPROVED** [review]

**Comment by TheFakeOG** [comment]
Yeeee boi
